File: languages/go/02-core/file-io-basics/example/csv-like-reader.go
// This extra example extends file-io-basics with line parsing and validation.
// Example purpose: read simple name,score rows from a text file.

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type studentScore struct {
	name  string
	score int
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("CSV-like input file path (name,score): ")
	path, err := reader.ReadString('\n')
	if err != nil && len(path) == 0 {
		fmt.Println("Could not read the file path.")
		return
	}
	path = strings.TrimSpace(path)

	handle, openErr := os.Open(path)
	if openErr != nil {
		fmt.Printf("Could not open file: %s\n", path)
		return
	}
	defer handle.Close()

	rows := make([]studentScore, 0)
	scanner := bufio.NewScanner(handle)
	lineNumber := 0

	for scanner.Scan() {
		lineNumber++
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		parts := strings.SplitN(line, ",", 2)
		if len(parts) != 2 {
			fmt.Printf("Skipping malformed line %d: %s\n", lineNumber, line)
			continue
		}

		score, parseErr := strconv.Atoi(strings.TrimSpace(parts[1]))
		if parseErr != nil || score < 0 || score > 100 {
			fmt.Printf("Skipping invalid score on line %d: %s\n", lineNumber, parts[1])
			continue
		}

		rows = append(rows, studentScore{name: parts[0], score: score})
	}

	fmt.Printf("\nValid rows: %d\n", len(rows))
	for _, row := range rows {
		fmt.Printf("- %s: %d\n", row.name, row.score)
	}
}

File: languages/cpp/02-core/file-io-basics/example/csv-like-reader.cpp
// This helper example focuses on separating row parsing from the higher-level file workflow.

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct StudentScore {
    string name;
    int score;
};

int main() {
    string path;
    cout << "CSV-like input file path (name,score): ";
    getline(cin, path);

    ifstream input(path);
    if (!input) {
        cout << "Could not open file: " << path << '\n';
        return 0;
    }

    vector<StudentScore> rows;
    string line;
    int lineNumber = 0;

    while (getline(input, line)) {
        ++lineNumber;
        if (line.empty()) {
            continue;
        }

        istringstream row(line);
        string name;
        string scoreText;

        if (!getline(row, name, ',') || !getline(row, scoreText)) {
            cout << "Skipping malformed line " << lineNumber << ": " << line << '\n';
            continue;
        }

        istringstream scoreStream(scoreText);
        int score = 0;
        if (!(scoreStream >> score) || score < 0 || score > 100) {
            cout << "Skipping invalid score on line " << lineNumber << ": " << scoreText << '\n';
            continue;
        }

        rows.push_back({name, score});
    }

    cout << "\nValid rows: " << rows.size() << '\n';
    for (const StudentScore& row : rows) {
        cout << "- " << row.name << ": " << row.score << '\n';
    }

    return 0;
}
